RFC-0040-nested-device-mesh.md

```diff
@@ -0,0 +1,123 @@
+# (Dynamic)NestedDeviceMesh
+
+**Authors:**
+* @Jackmin801
+* @samsja 
+* @JohannesHa
+
+## **Summary**
+The current abstraction for managing process groups in torch distributed is DeviceMesh, which provides a useful model for defining N-dimensional parallelisms.
+However, the current implementation has some limitations that make it unsuitable for dynamic heterogenous training workloads.
+The first limitation is that the DeviceMesh needs to be cuboid, assuming homogeneity among processes.
+This restriction limits its applicability for heterogeneous training workloads, where different devices might have varying workloads that do not map neatly onto a cuboid mesh.
+The second limitation is that the world size needs to be known at start time. This results in an inability to create dynamic training runs where groups of processes join and leave the training without restarting all the processes in the training run.
+
+To address the first limitation, we propose the introduction of a NestedDeviceMesh abstraction.
+This new abstraction allows for more flexibility by allowing nesting of DeviceMesh instances, an element of in a DeviceMesh can be a DeviceMesh, each potentially with different shapes and sizes.
+This would allow users to model more complex parallelism strategies, such as those required by heterogeneous workloads.
+
+To address the second limitation, we propose the process group creation API be rewritten to support dynamic world size by periodically polling the c10d store for the current world size and recreating the nccl communicators in the case of a change.
```

Comment:
I think this should be a separate abstraction. its relatively independent from the mesh part as we could have classic cuboic mesh that would require it

Reply:
yea. agree that they are somewhat orthogonal. included it in this rfc because i think its best to have the dynamic PGs before nested mesh and thus the nested mesh should consider it in design.

whether we will need a separate abstraction or can just modify the ProcessGroup abstraction im not so sure yet

Reply:
hrmm on second thought lets split it out. so two rfcs. one for the nested device mesh and one for dynamic PGs